### PR TITLE
Overhaul optimization tutorials

### PR DESCRIPTION
**Pull request by clayjohn**
These tutorials started out as a single brief batching tutorial written by lawnjelly and have turned into a complete overhaul of the optimization section of the tutorials. 

I did some substantial editing (including moving over the 3d optimizations tutorials and merging it in) and converted everything to rst. 

It would be great to have a few more sets of eyes to edit this as it is a large body of work.

The changes are only relevant to 3.2 right now as much of it is specific to batching and needs to link to the batching-specific ProjectSettings.

That being said, with not too much work, everything else can be ported into 4.0 with relative ease.

CC: @lawnjelly

### CODE REVIEW

File: tutorials/optimization/gpu_optimization.rst [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+addition this means that state changes, material and texture changes can be kept
+to a minimum.
+
+For more information on 2D batching see :ref:`doc_batching`.
+
+3D batching
+~~~~~~~~~~~
+
+In 3d we still aim to minimize drawcalls and state changes, however, it can be
+more difficult to batch together several objects into a separate drawcall. 3d
```
Comment (lawnjelly) [inline tutorials/optimization/gpu_optimization.rst:59]: `single drawcall` might be better.

File: tutorials/optimization/gpu_optimization.rst [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+=======================
+
+In particular GPUs on mobile devices can work in dramatically different ways to
+GPUs on desktop. Most mobiles use tile renderers. They split up the screen into
+regular sized tiles. This enables them to fit the entire tile into super fast
+cache memory, and reduce the reads and writes to main memory.
+
+There are some downsides though, it can make certain techniques much more
+complicated and expensive to perform. Tiles that rely on the results of
+rendering in different tiles can be very slow, or tiles that rely on the results
```
Comment (lawnjelly) [inline tutorials/optimization/gpu_optimization.rst:266]: `rendering in another tile can be very slow` might be better?

**Comment by clayjohn** [comment]
Github was being funny here. I pushed my changes yesterday and it updated in the commit, but the PR didn't update. I have force pushed again and now both are updated.

File: tutorials/optimization/cpu_optimization.rst [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+You can use a number of third party profilers to do this including Valgrind,
+VerySleepy, Visual Studio and Intel VTune. 
+
+.. note:: You may need to compile Godot from source in order to use a third
+          party profiler so that you have program database information
+          available. You can also use a debug build, however, note that the
+          results of profiling a debug build will be different to a release
+          build, because debug builds are less optimized. Bottlenecks are often
+          in a different place in debug builds, so you should profile release
+          builds when possible.
```
Comment (lawnjelly) [inline tutorials/optimization/cpu_optimization.rst:61]: wherever possible.

File: tutorials/optimization/cpu_optimization.rst [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+its children (Inclusive), the percentage of time spent within the function
+itself, excluding child functions (Self), the number of times the function is
+called, the function name, and the file or module.
+
+In this example we can see nearly all time is spent under the
+`Main::iteration()` function, this is the master function in the Godot source
+code that is called repeatedly, and causes frames to be drawn, physics ticks to
+be simulated, and nodes and scripts to be updated. A large proportion of the
+time is spent in the functions to render a canvas (66%), because in this example
+I was running a 2d benchmark. Below this we see that almost 50% of the time is
```
Comment (lawnjelly) [inline tutorials/optimization/cpu_optimization.rst:77]: a 2d benchmark was being profiled

File: tutorials/optimization/cpu_optimization.rst [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+
+For more information on threads see :ref:`doc_using_multiple_threads`.
+
+SceneTree
+=========
+
+Although Nodes are an incredibly powerful and versatile concept, be aware that
+every node has a cost. Built in functions such as `_process()` and
+`_physics_process()` propagate through the tree. This housekeeping can lead to
+situations when you have very large numbers of nodes.
```
Comment (lawnjelly) [inline tutorials/optimization/cpu_optimization.rst:216]: pathological situations

File: tutorials/optimization/general_optimization.rst [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+Limitations
+~~~~~~~~~~~
+
+CPU Profilers are often the 'go to' method for measuring performance, however
+they don't always tell the whole story.
+
+- Bottlenecks are often on the GPU, `as a result` of instructions given by the
+  CPU
+- Spikes can occur in the Operating System processes (outside of Godot) `as a
+  result` of instructions used in Godot (for example dynamic allocation)
```
Comment (lawnjelly) [inline tutorials/optimization/general_optimization.rst:76]: dynamic memory allocation

File: tutorials/optimization/general_optimization.rst [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+is (by definition) undesirable, performant software is the result of performant
+design.
+
+Performant design
+~~~~~~~~~~~~~~~~~
+
+The danger with encouraging people to ignore optimization until necessary, is
+that it conveniently ignores that the most important time to consider
+performance is at the design stage, before a key has even hit a keyboard. If the
+design / algorithms of a program is inefficient, then no amount of polishing the
```
Comment (lawnjelly) [inline tutorials/optimization/general_optimization.rst:167]: are inefficient

File: tutorials/optimization/general_optimization.rst [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+calculations) are slowing things down.
+
+As with design you should concentrate your efforts first on making sure the
+algorithms and data structures are the best they can be. Data access should be
+local (to make best use of CPU cache), and it can often be better to use compact
+storage of data (again, always profile to test results). Often you can make use
+of precalculation to do heavy computation ahead of time (e.g. at level load, or
+loading precalculated data files).
+
+Once algorithms and data is good, you can often make small changes in routines
```
Comment (lawnjelly) [inline tutorials/optimization/general_optimization.rst:223]: are good

File: tutorials/optimization/gpu_optimization.rst [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+=================
+
+Introduction
+~~~~~~~~~~~~
+
+The demand for new graphics features and progress almost guarantees that you
+will encounter graphics bottlenecks. Some of these can be CPU side, for instance
+in calculations inside the Godot engine to prepare objects for rendering.
+Bottlenecks can also occur on the CPU in the graphics driver, which sorts
+instructions to pass to the GPU. And finally bottlenecks also occur on the GPU
```
Comment (lawnjelly) [inline tutorials/optimization/gpu_optimization.rst:13]: to pass to the GPU, and in the transfer of these instructions.

File: tutorials/optimization/gpu_optimization.rst [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+Vulkan). The communication and driver activity involved can be quite costly,
+especially in OpenGL. If we can provide these instructions in a way that is
+preferred by the driver and GPU, we can greatly increase performance.
+
+Nearly every API command in OpenGL requires a certain amount of validation, to
+make sure the GPU is in the correct state. Even seemingly simple commands can
+lead to a flurry of behind the scenes housekeeping. Therefore the name of the
+game is reduce these instructions to a bare minimum, and group together similar
+objects as much as possible so they can be rendered together, or with the
+minimum number of state changes.
```
Comment (lawnjelly) [inline tutorials/optimization/gpu_optimization.rst:41]: expensive state changes

File: tutorials/optimization/gpu_optimization.rst [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+alternative is to join meshes ahead of time (static meshes in relation to each
+other). This can either be done by artists, or programmatically within Godot.
+
+There is also a cost to batching together objects in 3d. Several objects
+rendered as one cannot be individually culled. An entire city that is off screen
+will still be rendered if it is joined to a single blade of grass that is on
+screen. So attempting to batch together 3d objects should take account of their
+location and effect on culling. Often though the benefits of joining static
+objects will outweigh other considerations, especially for large numbers of low
+poly objects.
```
Comment (lawnjelly) [inline tutorials/optimization/gpu_optimization.rst:72]: Despite this, the benefits of joining static objects will often outweigh other considerations

File: tutorials/optimization/gpu_optimization.rst [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+dramatically over the years. Screen resolutions have increased (the area of a 4K
+screen is ``8,294,400`` pixels, versus ``307,200`` for an old ``640x480`` VGA
+screen, that is 27x the area), but also the complexity of fragment shaders has
+exploded. Physically based rendering requires complex calculations for each
+fragment.
+
+You can test whether a project is fill rate limited quite easily. Turn off vsync
+to prevent capping the frames per second, then compare the frames per second
+when running with a large window, to running with a postage stamp sized window.
+Usually you will find the fps increases quite a bit using a small window, which
```
Comment (lawnjelly) [inline tutorials/optimization/gpu_optimization.rst:166]: postage stamp sized window (you may also benefit from similarly reducing your shadow map size if using shadows)

File: tutorials/optimization/gpu_optimization.rst [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+
+As a warning, most Android devices do not support texture compression of
+textures with transparency (only opaque), so keep this in mind.
+
+Post processing / shadows
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Post processing effects and shadows can also be expensive in terms of fragment
+shading activity. Always test the impact of these on different hardware.
+
```
Comment (lawnjelly) [inline tutorials/optimization/gpu_optimization.rst:209]: Add to end: Reducing the size of shadow maps can increase performance, both in terms of writing, and reading the maps.

File: tutorials/optimization/optimizing_3d_performance.rst [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+Walking around a town for example, you may only be able to see a few buildings
+in the street you are in, as well as the sky and a few birds flying overhead. As
+far as a naive renderer is concerned however, you can still see the entire town.
+It won't just render the buildings in front of you, it will render the street
+behind that, with the people on that street, the buildings behind that. You
+quickly end up in situations where you are attempting to render 10x, or 100x
+more than what is visible.
+
+Things aren't quite as bad as they seem, because the Z-buffer usually only
+allows the GPU to fully shade the objects that are at the front. However,
```
Comment (lawnjelly) [inline tutorials/optimization/optimizing_3d_performance.rst:29]: z-buffer usually allows the GPU to fully shade only the objects that are at the front

File: tutorials/optimization/optimizing_3d_performance.rst [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+be available in core Godot in the future.
+
+Transparent objects
+~~~~~~~~~~~~~~~~~~~
+
+Godot sorts objects by :ref:`Material <class_Material>` and :ref:`Shader
+<class_Shader>` to improve performance. This, however, can not be done with
+transparent objects. Transparent objects are rendered from back to front to make
+blending with what is behind work. As a result, to use as few transparent
+objects as possible. If an object has a small section with transparency, try to
```
Comment (lawnjelly) [inline tutorials/optimization/optimizing_3d_performance.rst:61]: try to use

**Comment by lawnjelly** [comment]
It mostly seems okay, but I've read it so many times now I won't be able to pick up on problems. It would be good to have more input from more contributors. But I guess it can do as a starting point, then can be refined.

It could probably also benefit from more diagrams, even simple ones showing exactly the same as the text can help break things up and make it less of a slog to read! :blush:

File: tutorials/optimization/cpu_optimization.rst (hunk 5 above)
Comment (clayjohn) [reply tutorials/optimization/cpu_optimization.rst:216]: Pathological is a big word for a lot of users and will be hard to translate.

So I went with "This housekeeping can reduce performance when you have very large numbers of nodes."

**Comment by clayjohn** [comment]
@lawnjelly I totally agree.

**Comment by akien-mga** [comment]
Since a lot of it is relevant for `master`, I think it would be better to merge in the `master` branch, and it will end up in `3.2` too as I regularly merge `master` onto `3.2`: https://github.com/godotengine/godot-docs/commit/d48ff2b0a11a40280d716b6a95aed3d5e1ac2ed5

Then 4.0-specific changes can be done in the `master` branch where relevant, and I won't merge those in `3.2`.

**Comment by clayjohn** [comment]
@Akien won't I need to create two versions then? 4.0 doesn't have batching. Also VisualServer and Spatial material have been renamed.

**Comment by akien-mga** [comment]
The `master` branch is currently 95% documentation for Godot 3.2. There's only a few additions here and there for 4.0-specific changes, but we haven't propagated many of the 4.0 changes like node renames yet, or any of the rendering or GDScript changes. For example the `master` branch still has the page about GLES2 vs GLES3 differences. As a lot of refactoring will be done for #3390, and this should probably be made on the current post-3.2 state instead of in the middle of a massive documentation rewrite for Godot 4.0. I'm keeping the classref intentionally outdated so that we don't need to do node renames all over the docs.

Apart from the batching-specific documentation and the node renames, most of this seems applicable to future versions of Godot, so IMO it makes sense to merge in `master` as "this was the doc as of `3.2` that will be improved further for new versions".

But if you prefer merging in `3.2` and then making a new PR for `master`, that works too. I just fear that it will take a while before the rendering API for 4.0 is stable enough to have clear optimization docs, and having a section in `3.2` but missing from `master` might be weird. (Since our development philosophy is typically that `master` always has the most content.)

**Comment by clayjohn** [comment]
@akien-mga Thanks for the explanation. You have me convinced :) 

Some Nodes have been renamed and I am currently using the 3.2.2 names. Should I update this PR with the 4.0 names, or would you prefer to merge as is and then I can update the names in master in another PR?

**Comment by akien-mga** [comment]
Merging as is is good, as then I can backport it easily to `3.2` with a simple `git merge master`.

**Comment by akien-mga** [comment]
Actually, `master` does have some changes already which would be problematic, like the `StandardMaterial`/`SpatialMaterial3D` rename.

Let's stick to your initial plan then and merge this version in `3.2` first, and then you can do a near-identical forward port for `master` (with the few relevant references changed to avoid having Sphinx warnings on invalid references).

**Comment by clayjohn** [comment]
@akien-mga Sounds good! I am happy if this gets merged soon, or if we leave it a bit to get some more reviews. In a couple days I will have time to do another editing pass as well.

**Comment by akien-mga** [comment]
I didn't find/take time to review in depth, but as you were two to work on this, I think it's fairly safe to merge as is. We'll do adjustments as needed if we find typos or unclear sections when it's live.

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**